File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/StructStartSeekingRewriter.java
package com.codepoetics.raffia.streaming.rewriting;

import com.codepoetics.raffia.operations.Updater;
import com.codepoetics.raffia.paths.Path;
import com.codepoetics.raffia.streaming.EndOfLineFilter;
import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

public final class StructStartSeekingRewriter<T extends BasketWriter<T>> extends StreamingRewriter<T> {

  private final Path path;
  private final Updater updater;

  public StructStartSeekingRewriter(T target, Path path, Updater updater) {
    super(target, null, updater);
    this.path = path;
    this.updater = updater;
  }

  @Override
  public FilteringWriter<T> beginObject() {
    return StreamingRewriter.startObject(target, path, this, updater);
  }

  @Override
  public FilteringWriter<T> beginArray() {
    return StreamingRewriter.startArray(target, path, this, updater);
  }

  @Override
  public FilteringWriter<T> end() {
    throw new IllegalStateException("end() called when not writing object or array");
  }

  @Override
  public FilteringWriter<T> key(String key) {
    throw new IllegalStateException("key() called when not writing object");
  }

  @Override
  public FilteringWriter<T> advance(T newTarget) {
    return new EndOfLineFilter<>(newTarget);
  }

  @Override
  public FilteringWriter<T> add(String value) {
    return advance(target.add(value));
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    return advance(target.add(value));
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    return advance(target.add(value));
  }

  @Override
  public FilteringWriter<T> addNull() {
    return advance(target.addNull());
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/StructStartSeekingRewriter.kt
package com.codepoetics.raffia.streaming.rewriting

import com.codepoetics.raffia.java.api.Updater
import com.codepoetics.raffia.paths.Path
import com.codepoetics.raffia.streaming.EndOfLineFilter
import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

class StructStartSeekingRewriter<T : BasketWriter<T>>(target: T, private val path: Path, updater: Updater) : StreamingRewriter<T>(target, null, updater) {

    override fun beginObject(): FilteringWriter<T> {
        return StreamingRewriter.startObject(target, path, this, updater!!)
    }

    override fun beginArray(): FilteringWriter<T> {
        return StreamingRewriter.startArray(target, path, this, updater!!)
    }

    override fun end(): FilteringWriter<T> {
        throw IllegalStateException("end() called when not writing object or array")
    }

    override fun key(key: String): FilteringWriter<T> {
        throw IllegalStateException("key() called when not writing object")
    }

    override fun advance(newTarget: T): FilteringWriter<T> {
        return EndOfLineFilter(newTarget)
    }

    override fun add(value: String): FilteringWriter<T> {
        return advance(target.add(value))
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        return advance(target.add(value))
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        return advance(target.add(value))
    }

    override fun addNull(): FilteringWriter<T> {
        return advance(target.addNull())
    }
}
